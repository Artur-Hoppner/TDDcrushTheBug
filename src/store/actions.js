const actions = {
  addThisToCart(context, product) {
    context.commit('addToCart', product);
  },
  changeThisProductButtonToggle(context, boolean) {
    context.commit('changeProductButtonToggle', boolean);
  },
  removeThisFromCart(context, id) {
    context.commit('removeFromCart', id);
  },
  increaseThisQuantity(context, product) {
    context.commit('increaseQuantity', product);
  },
  decreaseThisQuantity(context, product) {
    context.commit('decreaseQuantity', product);
  },
  getByThisKeyword(context, keyWord) {
    context.commit('getByKeyword', keyWord);
  },
  displayThisEntireArray(context) {
    context.commit('displayEntireArray');
  }
};

export default actions;
