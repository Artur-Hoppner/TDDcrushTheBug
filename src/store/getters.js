//***********************/
//*** GLOBAL GETTERS ***/
//*********************/

const getters = {
  getProductButtonToggle: state => {
    return state.productButtonToggle;
  },
  getCartProducts: state => {
    return state.cart;
  },
  filteredByKeyword: state => {
    return state.filteredProducts;
  },
  getOrderInfo: state => {
    return state.orderInfo;
  },
  getOrderItems: state => {
    return state.orderInfo.orderList;
  },
  cartIconItems: state => {
    let total = 0;
    state.cart.forEach(cart => {
      total += cart.quantity;
    });
    return total;
  },
  showHamburger: state => {
    return state.isBurgerActive;
  }
};

export default getters;
