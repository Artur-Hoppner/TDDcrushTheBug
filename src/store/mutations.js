//*************************/
//*** GLOBAL MUTATIONS ***/
//***********************/

const mutations = {
  //*************************************/
  //*** CART ARRAY RELATED MUTATIONS ***/
  //***********************************/
  //Add product to cart, takes product object as argument
  addToCart(state, product) {
    //Check if cart already has existing item so no doublettes will be created, only update quantity & price
    let checkExisting = state.cart.filter(item => item.id == product.id);
    if (checkExisting.length >= 1) {
      let index = state.cart.findIndex(item => item.id === product.id);
      state.cart[index].quantity++;
      state.cart[index].price += product.price;
    } else {
      state.cart.push({
        id: product.id,
        tag: product.tag,
        title: product.title,
        desc: product.desc,
        price: product.price,
        quantity: 1
      });
    }
  },
  //Remove Product from cart using index
  removeFromCart(state, id) {
    let findIndex = state.cart.find(item => item.id == id);
    let index = state.cart.indexOf(findIndex);
    state.cart.splice(index, 1);
  },
  //Increase Product Quantity in cart
  increaseQuantity(state, product) {
    //Check for default values of object in original allProducts array
    //Increase Product quantity and price
    let adjustPrice = state.allProducts.find(item => item.id == product.id);
    let index = state.cart.findIndex(item => item.id === product.id);
    state.cart[index].quantity++;
    state.cart[index].price += adjustPrice.price;
  },
  //Decrease Product Quantity in cart
  decreaseQuantity(state, product) {
    //Check for default values of object in original allProducts array
    //Decrease Product quantity and price
    let adjustPrice = state.allProducts.find(item => item.id == product.id);
    let index = state.cart.findIndex(item => item.id === product.id);
    state.cart[index].quantity--;
    state.cart[index].price -= adjustPrice.price;
    if (state.cart[index].quantity == 0) {
      state.cart.splice(index, 1);
    }
  },
  //Create order object based on state.cart
  checkoutCreateOrder(state) {
    if (state.cart != '') {
      //Randomize ordernumber and calculate total cost
      let newOrderNumber = Math.floor(Math.random() * 100000 + 1);
      let totalCost = 0;
      for (let i = 0; i < state.cart.length; i++) {
        totalCost += state.cart[i].price;
      }
      //Create Order object
      state.orderInfo = {
        orderNumber: newOrderNumber,
        orderCost: totalCost,
        deliveryETA: '3',
        orderList: state.cart
      };
    } else {
      return false;
    }
  },
  //*************************************************/
  //*** FILTEREDPRODUCTS ARRAY RELATED MUTATIONS ***/
  //***********************************************/
  //Take input searchword as argument, then use it to filter through allproducts array and match all products including the searchword
  getByKeyword(state, keyWord) {
    const search = keyWord.toLowerCase();
    if (!search.length) {
      state.filteredProducts = state.allProducts;
    } else {
      let foundProducts = [];
      foundProducts = state.allProducts.filter(item => item.tag.match(search));
      state.filteredProducts = foundProducts;
    }
  },
  //Reset the displayArray so it shows all products
  displayEntireArray(state) {
    state.filteredProducts = state.allProducts;
  },
  //*********************************/
  //*** TOGGLE RELATED MUTATIONS ***/
  //*******************************/
  //Change the toggle which decides if ProductItem shows AddToCart button or Increment, Decrement & RemoveFromCart depending on if we are viewing cart or shopping page
  changeProductButtonToggle(state, boolean) {
    state.productButtonToggle = boolean;
  },
  changeCart(state) {
    state.showCart = !state.showCart;
  },
  toggle(state) {
    state.isBurgerActive = !state.isBurgerActive
  }
};

export default mutations;
