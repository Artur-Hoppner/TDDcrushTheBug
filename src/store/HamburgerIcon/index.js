import mutations from './mutations.js';
import getters from './getters.js';
import actions from './actions.js';

const state = {
  namespaced: true,

  state: {
    isBurgerActive: false
  },
  mutations: mutations,
  getters: getters,
  actions: actions
};

export default state;
