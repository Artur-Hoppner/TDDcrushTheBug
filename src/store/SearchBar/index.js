import mutations from './mutations.js';
import getters from './getters.js';
import actions from './actions.js';

const state = {
  namespaced: true,
  //TODO: fix placeholder array with actual products!
  state: {
    allProducts: [
      'cat',
      'dog',
      'parrot',
      'goldfish',
      'horse',
      'elephant',
      'ant',
      'snake',
      'crow',
      'cow',
      'eagle',
      'rhinoceros',
      'chimpanzee'
    ],
    filteredProducts: []
  },
  mutations: mutations,
  getters: getters,
  actions: actions
};

export default state;
