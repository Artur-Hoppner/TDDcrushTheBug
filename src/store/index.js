import Vue from 'vue';
import Vuex from 'vuex';
import searchbar from './SearchBar/index.js';
import actions from './actions.js';
import mutations from './mutations.js';
import getters from './SearchBar/getters.js';
import hamburgerIcon from './HamburgerIcon/index.js'

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    allProducts: [],
    cart: [],
    searchedProducts: []
  },
  actions: actions,
  mutations: mutations,
  getters: getters,
  modules: {
    searchbar: searchbar,
    hamburgerIcon: hamburgerIcon
  }
});
