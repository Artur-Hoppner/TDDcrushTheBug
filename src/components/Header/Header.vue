<template>
  <nav class="header">
    <HamburgerIcon />
  </nav>
</template>

<script>
import HamburgerIcon from '@/components/HamburgerIcon/HamburgerIcon.vue';
//TODO: import cartIcon when it's finished

export default {
  name: 'Header',
  components: {
    HamburgerIcon
  }
};
</script>

<style></style>
