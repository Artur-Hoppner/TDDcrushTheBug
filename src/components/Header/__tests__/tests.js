import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import store from '../../../store/index.js';
import Header from '@/components/Header/Header.vue';

//create a local instance of our vue
const localVue = createLocalVue();
//make our instance use vuex
localVue.use(Vuex);

describe('User loads the page', () => {
  test('Test so that everything renders correctly', async () => {
    //Mount component with store & localvue, take snapshot and check so it renders correctly
    const wrapper = shallowMount(Header, { store, localVue });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('Test if method goTo is called on #icon click', async () => {
    // Arrange
    const goTo = jest.fn();
    const wrapper = shallowMount(Header, {
      localVue,
      methods: {
        goTo
      }
    });
    // Act
    const input = wrapper.find('#icon');
    console.log(input, 'input data');
    await input.trigger('click');
    await localVue.nextTick();
    // Assert
    expect(goTo).toHaveBeenCalled();
  });
});
