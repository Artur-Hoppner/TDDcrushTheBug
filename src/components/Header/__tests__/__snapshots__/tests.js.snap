// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User loads the page Test so that everything renders correctly 1`] = `
<div
  class="wrapper"
>
  <nav
    class="header"
  >
    <img
      alt=""
      class="header-icon"
      src="@/assets/bug-icon.png"
    />
     
    <h1
      id="text"
    >
      Crush The Bug
    </h1>
     
    <section
      class="icons"
    >
      <carticon-stub />
       
      <hamburgericon-stub />
    </section>
  </nav>
</div>
`;
